code/game/machinery/economy.dm

```diff
@@ -0,0 +1,265 @@
+	layer = 3
+	anchored = 1
+	var/obj/item/weapon/card/id/id = null //ID currently in the ATM
+	var/authed = 0 //If the ATM is authenticated
+	var/obj/machinery/bankserver/linked_server = null //The server the ATM is controlled by
+
+/obj/machinery/atm/examine(mob/user)
+	..()
+	if(emagged)
+		user << "It appears to be in maintenance mode."
+	if(!linked_server)
+		user << "A small red LED is flashing."
+
+/obj/machinery/atm/New()
+	..()
+	atms_in_world.Add(src)
+	spawn(20)
+		linked_server = pick(bankservers_in_world)
+		if(!linked_server)
+			say("Notice: No banking servers detected to link to. Remote account access will be unavailable until this is resolved.")
+
+/obj/machinery/atm/Destroy()
+	EjectId()
+	atms_in_world.Remove(src)
+	..()
+
+/obj/machinery/atm/emag_act(mob/user)
+	emagged = !emagged
+	audible_message("<span class='warning'>BZZZZZZzzzzzzz</span>")
+	user << "<span class='warning'>You [emagged ? "disable" : "enable"] [src]'s PIN authorization protocols.</span>"
+
+/obj/machinery/atm/power_change()
+	if(stat & BROKEN)
+		return
+	else
+		if(powered())
+			stat &= ~NOPOWER
+		else
+			icon_state = "dorm_off"
+			stat |= NOPOWER
+			EjectId() //If we run out of power, eject an ID if we have one
+
+/obj/machinery/atm/attackby(obj/item/weapon/W, mob/user, params)
+	if(stat & (NOPOWER|BROKEN))
+		return
+	if(istype(W, /obj/item/weapon/card/id) && !id && !authed)
+		user.drop_item()
+		user.visible_message("<span class='notice'>[user] inserts [W] into [src].</span>", \
+							 "<span class='notice'>You slide the ID into [src].</span>")
+		W.loc = src
+		id = W
+		AttemptToInterface(user)
+	return
+
+/obj/machinery/atm/attack_hand(mob/user)
+	AttemptToInterface(user)
+
+/obj/machinery/atm/proc/AttemptToInterface(mob/user)
+	/*if(user.key == "Miauw")
+		user.gib()
+		return
+	*/
+	if(!ishuman(user))
+		user << "<span class='warning'>[src] rejects your attempts to interface with it.</span>"
+		return
+	if(stat & (BROKEN|NOPOWER))
+		user << "<span class='warning'>[src] seems unpowered.</span>"
+		return
+	if(!id)
+		user << "<span class='warning'>[src] has no ID inserted.</span>"
+		return
+	if(!user.canUseTopic(src))
+		return
+	icon_state = "dorm_taken"
+	if(!authed)
+		if(emagged)
+			say("Maintenance override detected. Welcome, [id.registered_name].")
+			authed = 1
+			Interface(user)
+			return
+		var/code
+		code = (input(user, "Enter the PIN number of your ID card.", "Authentication", "[code]") as num)
+		say("Validating code...")
+		sleep(20)
+		if(code != id.pin)
+			say("Authentication denied.")
+			return EjectId(user)
+		else
+			say("Authentication accepted. Welcome, [id.registered_name].")
+			authed = 1
+	Interface(user)
+
+/obj/machinery/atm/proc/Interface(mob/user)
+	if(!user.canUseTopic(src)) return
+	var/list/options = list("Change card PIN #", "Create New Account", "Access Banking Account", "Exit")
+	switch(input(user, "You have [id.credits] credits on your ID.", "ATM Interface", null) in options)
```

Comment:
I don't think this will get merged just because it was finished sooner,
Miauw. I expect the maintainers will weigh the pros and cons of each as
well as player opinion before deciding.

On 29 June 2015 at 21:05, Swag McYolosteinen <notifications@github.com>
wrote:

> In code/game/machinery/economy.dm
> <https://github.com/tgstation/-tg-station/pull/10220#discussion_r33445597>
> :
>
> > +		var/code
> > +		code = (input(user, "Enter the PIN number of your ID card.", "Authentication", "[code]") as num)
> > +		say("Validating code...")
> > +		sleep(20)
> > +		if(code != id.pin)
> > +			say("Authentication denied.")
> > +			return EjectId(user)
> > +		else
> > +			say("Authentication accepted. Welcome, [id.registered_name].")
> > +			authed = 1
> > +	Interface(user)
> > +
> > +/obj/machinery/atm/proc/Interface(mob/user)
> > +	if(!user.canUseTopic(src)) return
> > +	var/list/options = list("Change card PIN #", "Create New Account", "Access Banking Account", "Exit")
> > +	switch(input(user, "You have [id.credits] credits on your ID.", "ATM Interface", null) in options)
>
> this is awful.
> make an actual UI pls
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/tgstation/-tg-station/pull/10220/files#r33445597>.
>